Support for memory-backed file descriptor

dalehamel opened the pull request:
This abstracts the creation of the temporary ELF file to allow for two possible approaches:

- The current (default) approach of allocating a temporary file to `/tmp` and dlopen'ing it 
- A new (optional) approach, to use a memory-backed file descriptor, if `LIBSTAPSDT_MEMORY_BACKED_FD` is specified.

The new approach his beneficial for two reasons:

- There is no need to clean up the file, it is automatically removed when the process dies (the primary reason for this approach).
- No calls are made to the underlying filesystem, everything is done in memory. While tmpfs is probably already memory-based, this could be advantageous in some circumstances.

libusdt does something similar, where it just [valloc's directly into the processes address space](https://github.com/chrisa/libusdt/blob/master/usdt_probe.c#L99), so there is no need for a temporary file to clean up.

Overall, I find that doing tho whole thing in memory and having it be automatically cleaned up with the process's lifecycle is a bit more elegant.

The drawbacks to this approach are:

- ~It is not yet supported by bcc, but I've submitted https://github.com/iovisor/bcc/pull/2314 to show how this could be utilized.~ Support has been added to BCC by https://github.com/iovisor/bcc/pull/2314
- If a large number of providers are allocated, the process will have a large number of file descriptors open. This is probably not a problem in practice.
- Support for `memfd_create` was not added until linux 3.17 (not a problem in practice, as BCC needs 4.1+ to do anything meaningful with eBPF) and glibc 2.27. The system call may not be defined on some systems.

I think the drawbacks are mitigated by hiding it behind a preprocessor macro check, as the behavior remains unchanged. In the future, hopefully once https://github.com/iovisor/bcc/pull/2314 is in a major bcc release, we could remove the macro and default to a memory-based shared object, falling back to temporary file if we detect that `memfd_create` is not supported.

Attaching a probe (`global:helloworld`) this way, the resulting process looks like:

/proc/${PID}/maps:

```
...
7f9e79f0c000-7f9e79f0d000 r-xp 00000000 00:05 11706507                   /memfd:libstapsdt:global (deleted)
7f9e79f0d000-7f9e7a10c000 ---p 00001000 00:05 11706507                   /memfd:libstapsdt:global (deleted)
7f9e7a10c000-7f9e7a10d000 rw-p 00000000 00:05 11706507                   /memfd:libstapsdt:global (deleted)
...
```

And if we check out the file descriptors for the process:

```
lrwx------ 1 dale.hamel dale.hamel 64 Apr 14 20:57 7 -> '/memfd:libstapsdt:global (deleted)'
```

So if we check for elf notes on that fd:

```
$readelf --notes /proc/${PID}/fd/7

Displaying notes found in: .note.stapsdt
  Owner                 Data size       Description
  stapsdt              0x00000039       NT_STAPSDT (SystemTap probe descriptors)
    Provider: global
    Name: helloword
    Location: 0x0000000000000260, Base: 0x0000000000000318, Semaphore: 0x0000000000000000
    Arguments: 8@%rdi -8@%rsi
```

And if I run tplist -p
```
$ tplist -p ${PID} | grep global
/proc/12646/fd/7 global:hello_nsec
```

Or use my latest branch of bpftrace with wildcard USDT support:
```
$ bpftrace -l 'usdt:*:global:*' -p ${PID}
usdt:/proc/12646/fd/7:global:helloworld
```

I can attach to the probe by that path, or by PID (using my bcc branch)

If I disable the provider, the file descriptor is closed and the elf file is removed from the memory map.

dalehamel committed:
Support for memory-backed file description

dalehamel commented on src/libstapsdt.c:
```diff
@@ -18,6 +19,19 @@
+
+// Note that linux must be 3.17 or greater to support this
+static inline int memfd_create(const char *name, unsigned int flags) {
+  return syscall(__NR_memfd_create, name, flags);
```
should probbaly check that `__NR_memfd_create` is defined or else bail out elegantly.

dalehamel commented:
I've marked this as ready for review, as the dependent patch has landed in BCC.

mmarchini commented:
Awesome! I was not aware of `memfd`, and this seems like a much better approach compared to creating a file in the filesystem. `libusdt` doesn't have to write to a file first because Solaris/DTrace have an interface to register User Tracepoints.

IMO memfd should be the default, and we should fall back to tmpfs only if necessary. Also, it might be useful to have a way to choose which filesystem to use during runtime?

dalehamel commented:
@mmarchini great - i'll remove the macro then, and make it the default, falling back only if we can't find the necessary call.

> Also, it might be useful to have a way to choose which filesystem to use during runtime?

to do this I'll need to add a new method signature for providerLoad probably, to force using /tmp files. If the original interface is called, it will go through the fallback path.

I'll modify this PR accordingly :)

mmarchini commented:
@dalehamel did you had time to look into this?

javierhonduco commented:
I've tried this PR on our internal Python wrapper and seems to work fine, It passes all our integration tests! 🎉 

I was curious about performance and it seems to be the same as before. Intuitively I thought the performance would be better now. Just in case, will double check the simple benchmarks I wrote in the next couple of days

dalehamel committed:
Use memfd by default, allow fall back to tmpfile

This Makes the default behavior of libstapsdt to try and use a memory-backed
file descriptor for systems that support this syscall.

At runtime, this behavior can be changed with the function providerUseMemfd

dalehamel commented:
@mmarchini ok sorry for the super long delay here, I have made the following changes:

- The tests are run on both trusty and bionic. Trusty doesn't support memfd, so uses the old codepath (glibc too old). bionic does, and tests the new codepath.
- Support for memfd is detected automatically at compile time, if available it's used by default
- A new function is added to the public interface, to allow toggling the behavior of using memfd or no, on systems that support it.

I also found some memory leaks that I think I missed on the last pass / weren't running in the test suite before. I've fixed those - the valgrind test is really helpful :)

Can you please take another look?

dalehamel commented on src/libstapsdt.h:
```diff
@@ -47,10 +47,20 @@ typedef struct SDTProvider {
+runtime by calling this with use_memfd = 0 prior after providerInit, and before
+providerLoad.
+*/
+int providerUseMemfd(SDTProvider_t *provider, const int use_memfd);
```
because it is pure C, we can't overload the providerInit function like I'd normally prefer to, so I added a helper instead.

dalehamel commented on tests/test-errors.c:
```diff
@@ -11,7 +18,18 @@ int testElfCreationError() {
   SDTProvider_t *provider;
   SDTError_t errno;
+#ifdef HAVE_LIBSTAPSDT_MEMORY_BACKED_FD
+  char *providerLongName = "test/probe/creation/error/with/a/name/longer/than/"
```
the path name used to fail the test for tempfiles is actually valid for memfd, so we force it to fail by sending it a name that is too long.

javierhonduco commented on src/libstapsdt.c:
```diff
@@ -80,30 +115,63 @@ SDTProbe_t *providerAddProbe(SDTProvider_t *provider, const char *name, int argC
+    }
+
+    snprintf(path_buffer, sizeof(path_buffer), "/proc/%d/fd/%d", getpid(), *fd);
+    filename = calloc(sizeof(char), strlen(path_buffer) + 1);
```
Should we check if `calloc` returns `NULL`? (we may want to also check the other allocation callsites that predate this PR)

mmarchini replied:
idr why I didn't originally check if `calloc` returns `NULL` on other locations. We probably should yes, and that can be done in a separate PR.

javierhonduco commented on src/libstapsdt.c:
```diff
@@ -80,30 +115,63 @@ SDTProbe_t *providerAddProbe(SDTProvider_t *provider, const char *name, int argC
+    return filename;
+  }
+#endif
+  filename = calloc(sizeof(char), strlen("/tmp/-XXXXXX.so") + strlen(name) + 1);
```
Same as above

javierhonduco reviewed (approved):
Looks good to me!

mmarchini commented:
Great! I might take a few weeks to get to it though. Just got back from vacation and I'm on a few conferences now, so my backlog is quite huge. Will try to prioritize it though, as it's a nice addition :)

dalehamel commented:
> Great! I might take a few weeks to get to it though. Just got back from vacation and I'm on a few conferences now, so my backlog is quite huge

No worries @mmarchini hope you had a restful vacation!

dalehamel commented:
Hi @mmarchini a friendly ping, in case this has fallen off your radar and you can find some time to take a look

javierhonduco commented:
Hi! It would be great if this PR could be merged. At my workplace we currently have this custom patch applied but it would be great if we don't depend on custom patches 😄

mmarchini commented on src/libstapsdt.h:
```diff
@@ -47,10 +47,20 @@ typedef struct SDTProvider {
+using a memory-backed file descriptor. This behavior can be overridden at
+runtime by calling this with use_memfd = 0 prior after providerInit, and before
+providerLoad.
+*/
```
Let's provide an `enum` here instead of `0`/`1`, so it's less confusing for users.

dalehamel replied:
I believe I've addressed this in https://github.com/dalehamel/libstapsdt/commit/4c62f850757d7ba046e98809fc94df39547613a8 please take a look and let me know what you think

mmarchini commented on src/libstapsdt.c:
```diff
@@ -18,6 +20,21 @@
+
+#ifdef __NR_memfd_create // older glibc may not have this syscall defined
+#define HAVE_LIBSTAPSDT_MEMORY_BACKED_FD
+#define F_SEAL_SEAL 0x0001 /* prevent further seals from being set */
```
Can we `#include <linux/fcntl.h>` instead (see [this SO answer](https://stackoverflow.com/a/39353581))? Or are there any caveats of doing that?

dalehamel replied:
Good catch, I've made the necessary change in https://github.com/dalehamel/libstapsdt/commit/c4248d0c2f5d0bcff1f9359548024bc1a4b80e9b

mmarchini reviewed (approved):
LGTM with the `enum` added. I also think the file opening/closing code should be moved to a separate file, but we can do that in a separate PR.

I think there is a bug if the user tries to start two providers with the same name (since they'll have the same filename), so we might want to handle that too (again, can be done in a follow up).

I forked the project, it's now located at https://github.com/linux-usdt/libstapsdt and I'll be continuing development there. Feel free to open the PR there if you're still interested in working on it. If not, I'm happy to make the enum changes and cherry-pick the commits.

Last but not least, I do wonder if this should result in `libstapsdt2`, because it's technically a breaking change (more from the POV of the consumers than the programs running it though, since now older bcc/bpftrace versions won't be able to load the probes).

dalehamel committed:
Introduce MemFD_Option_t enum for controlling usage of memfd

This introduces a new enumerated type, MemFD_Option_t, which is used in lieu
of the previous convention of using just an integer type.

All references to these values have been updated, and the symbolic values of
"memfd_enabled" is used in place of the literal "1", and "memfd_disabled" is
used in place of the literel "0".

dalehamel committed:
Use linux/fcntl for F_SEAL_SEAL constant

This avoids duplicating the definition of the constant F_SEAL_SEAL, and
allows relying on an existing standard linux header for it.
